bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * BGP Conditional advertisement
+ * Copyright (C) 2020  Samsung Research Institute Bangalore.
+ *			Madhurilatha Kuruganti
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "bgpd/bgp_conditional_adv.h"
+
+const char *get_afi_safi_str(afi_t afi, safi_t safi, bool for_json);
+
+static route_map_result_t
+bgp_check_rmap_prefixes_in_bgp_table(struct bgp_table *table,
+				     struct route_map *rmap)
+{
+	struct bgp_dest *dest;
+	struct attr dummy_attr;
+	struct bgp_path_info path;
+	const struct prefix *dest_p;
+	struct bgp_path_info *pi;
+	route_map_result_t ret = RMAP_PERMITMATCH;
+
+	for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+		dest_p = bgp_dest_get_prefix(dest);
+		if (!dest_p)
+			continue;
+
+		for (pi = bgp_dest_get_bgp_path_info(dest); pi; pi = pi->next) {
+			dummy_attr = *pi->attr;
+			path.peer = pi->peer;
+			path.attr = &dummy_attr;
+
+			ret = route_map_apply(rmap, dest_p, RMAP_BGP, &path);
+			if (ret == RMAP_PERMITMATCH)
```

Comment:
We must reset attributes it was before with `bgp_attr_flush(&dummy_attr)` if not RMAP_PERMITMATCH. Otherwise, it will lead to changed route-maps.

Reply:
If you don't mind I have a doubt in this comment.
as you said, in case of subgroup_announce_check() attributes are modified before route_map_apply() and when result is DENYMATCH, bgp_attr_flush() is called.
But here we are using the attr as it is right do we need bgp_attr_flush() ?